### PR TITLE
:sparkles: Add files support for Azure DevOps

### PR DESCRIPTION
**Pull request by JamieMagee**
#### What kind of change does this PR introduce?

Files support for the Azure DevOps client

- [x] PR title follows the guidelines defined in our [pull request documentation](https://github.com/ossf/scorecard/blob/main/CONTRIBUTING.md#pr-process)

#### What is the current behavior?

`ListFiles`, `LocalPath`, and `GetFileReader` all throw an unsupported feature error.

#### What is the new behavior (if this is a feature change)?**

`ListFiles`, `LocalPath`, and `GetFileReader` have implementations

- [ ] Tests for the changes have been added (for bug fixes/features)

#### Which issue(s) this PR fixes

Next step of #4177

#### Does this PR introduce a user-facing change?

For user-facing changes, please add a concise, human-readable release note to
the `release-note`

(In particular, describe what changes users might need to make in their
application as a result of this pull request.)

```release-note
Add files support to Azure DevOps client
```

### CODE REVIEW

File: clients/azuredevopsrepo/zip.go [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+	"github.com/microsoft/azure-devops-go-api/azuredevops/v7"
+
+	sce "github.com/ossf/scorecard/v5/errors"
+)
+
+const (
+	repoDir      = "repo*"
+	repoFilename = "azuredevopsrepo*.zip"
+	maxSize      = 100 * 1024 * 1024 // 100MB limit
+)
```
Comment (spencerschrock) [inline clients/azuredevopsrepo/zip.go:39]: Just to be clear, this is per file?

File: clients/azuredevopsrepo/zip.go [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+		return fmt.Errorf("zip.NewReader: %w", err)
+	}
+
+	z.files = make([]string, 0, len(zipReader.File))
+	for _, file := range zipReader.File {
+		//nolint:gosec // G305: Handling of file paths is done below
+		filenamepath := filepath.Join(z.tempDir, file.Name)
+		if !strings.HasPrefix(filepath.Clean(filenamepath), filepath.Clean(z.tempDir)) {
+			return errInvalidFilePath
+		}
```
Comment (spencerschrock) [inline clients/azuredevopsrepo/zip.go:183]: No action required, just posting for discussion / posterity.

Go takes a slightly different sanitization technique, but only with a godebug value set:
https://go.dev/doc/godebug#go-120
> Go 1.20 introduced support for rejecting insecure paths in tar and zip archives, controlled by the [tarinsecurepath setting](https://go.dev/pkg/archive/tar/#Reader.Next) and the [zipinsecurepath setting](https://go.dev/pkg/archive/zip/#NewReader). These default to tarinsecurepath=1 and zipinsecurepath=1, preserving the behavior of earlier versions of Go. A future version of Go may change the defaults to tarinsecurepath=0 and zipinsecurepath=0.

File: clients/azuredevopsrepo/zip.go [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+}
+
+func (z *zipHandler) cleanup() error {
+	if err := os.RemoveAll(z.tempDir); err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("os.Remove: %w", err)
+	}
+
+	z.files = nil
+	return nil
+}
```
Comment (spencerschrock) [inline clients/azuredevopsrepo/zip.go:251]: Does it make sense to call this from within the toplevel Client.Close()?
https://github.com/ossf/scorecard/blob/f4d721d307d53e54346aa165b796e5ecfa5460cb/clients/azuredevopsrepo/client.go#L185-L187

File: clients/azuredevopsrepo/zip.go [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+
+	stat, err := in.Stat()
+	if err != nil {
+		return fmt.Errorf("os.File.Stat: %w", err)
+	}
+
+	zipReader, err := zip.NewReader(in, stat.Size())
+	if err != nil {
+		return fmt.Errorf("zip.NewReader: %w", err)
+	}
```
Comment (spencerschrock) [inline clients/azuredevopsrepo/zip.go:175]: optional: [zip.OpenReader](https://pkg.go.dev/archive/zip#OpenReader) does a lot of this for you.
```go
r, err := zip.OpenReader(z.tempZipFile)
if err != nil {
	return fmt.Errorf("openning zip: %w", err)
}
defer r.Close()
```

**Review by spencerschrock — COMMENTED** [review]
At a high level this looks good and matches the style in the other ones, but it's also late on a Friday. Left a few comments. 

Out of curiosity, how do file-only based checks look like on your test repo now? Such as License, and Binary-Artifacts?

File: clients/azuredevopsrepo/zip.go (hunk 1 above)
Comment (JamieMagee) [reply clients/azuredevopsrepo/zip.go:39]: Yes. I added this to resolve the G110 from gosec:

> G110: Potential DoS vulnerability via decompression bomb

**Comment by JamieMagee** [comment]
> At a high level this looks good and matches the style in the other ones, but it's also late on a Friday. Left a few comments.

Thank you! And I completely understand. 

> Out of curiosity, how do file-only based checks look like on your test repo now? Such as License, and Binary-Artifacts?

They look good. Here's are a couple of examples

<details>
<summary>No license</summary>

```
$ env SCORECARD_EXPERIMENTAL=1 go run . --repo dev.azure.com/jamiemagee/jamiemagee/_git/jamiemagee --checks License
Starting [License]
Finished [License]

RESULTS
-------
Aggregate score: 0.0 / 10

Check scores:
|--------|---------|---------------------------|--------------------------------------------------------------------|
| SCORE  |  NAME   |          REASON           |                     DOCUMENTATION/REMEDIATION                      |
|--------|---------|---------------------------|--------------------------------------------------------------------|
| 0 / 10 | License | license file not detected | https://github.com/ossf/scorecard/blob/main/docs/checks.md#license |
|--------|---------|---------------------------|--------------------------------------------------------------------|
```

</details>

<details>
<summary>With license</summary>

```
$ env SCORECARD_EXPERIMENTAL=1 go run . --repo dev.azure.com/jamiemagee/jamiemagee/_git/jamiemagee --checks License --commit d26320a77a943045905aed76a7c08a353b3bc3c2
Starting [License]
Finished [License]

RESULTS
-------
Aggregate score: 9.0 / 10

Check scores:
|--------|---------|-----------------------|--------------------------------------------------------------------|
| SCORE  |  NAME   |        REASON         |                     DOCUMENTATION/REMEDIATION                      |
|--------|---------|-----------------------|--------------------------------------------------------------------|
| 9 / 10 | License | license file detected | https://github.com/ossf/scorecard/blob/main/docs/checks.md#license |
|--------|---------|-----------------------|--------------------------------------------------------------------|
```

</details>

<details>
<summary>No binary artifacts</summary>

```
env SCORECARD_EXPERIMENTAL=1 go run . --repo dev.azure.com/jamiemagee/jamiemagee/_git/jamiemagee --checks Binary-Artifacts --commit d26320a77a943045905aed76a7c08a353b3bc3c2
Starting [Binary-Artifacts]
Finished [Binary-Artifacts]

RESULTS
-------
Aggregate score: 10.0 / 10

Check scores:
|---------|------------------|-------------------------------|-----------------------------------------------------------------------------|
|  SCORE  |       NAME       |            REASON             |                          DOCUMENTATION/REMEDIATION                          |
|---------|------------------|-------------------------------|-----------------------------------------------------------------------------|
| 10 / 10 | Binary-Artifacts | no binaries found in the repo | https://github.com/ossf/scorecard/blob/main/docs/checks.md#binary-artifacts |
|---------|------------------|-------------------------------|-----------------------------------------------------------------------------|
```

</details>

<details>
<summary>With binary artifacts</summary>

```
env SCORECARD_EXPERIMENTAL=1 go run . --repo dev.azure.com/jamiemagee/jamiemagee/_git/jamiemagee --checks Binary-Artifacts --commit 696984c21bbf3f37cf4d4809e0b4ff1051c40f09
Starting [Binary-Artifacts]
Finished [Binary-Artifacts]

RESULTS
-------
Aggregate score: 9.0 / 10

Check scores:
|--------|------------------|--------------------------------|-----------------------------------------------------------------------------|
| SCORE  |       NAME       |             REASON             |                          DOCUMENTATION/REMEDIATION                          |
|--------|------------------|--------------------------------|-----------------------------------------------------------------------------|
| 9 / 10 | Binary-Artifacts | binaries present in source     | https://github.com/ossf/scorecard/blob/main/docs/checks.md#binary-artifacts |
|        |                  | code                           |                                                                             |
|--------|------------------|--------------------------------|-----------------------------------------------------------------------------|
```

</details>

File: clients/azuredevopsrepo/zip.go (hunk 2 above)
Comment (spencerschrock) [inline clients/azuredevopsrepo/zip.go:183]: I think the prefix check is missing the final path separator:
```go
filenamepath := filepath.Join(z.tempDir, file.Name)
if !strings.HasPrefix(filepath, filepath.Clean(z.tempDir) + string(os.PathSeparator)) {
	return errInvalidFilePath
}
```

Based on the existing version in the GitHub code:
https://github.com/ossf/scorecard/blob/f4d721d307d53e54346aa165b796e5ecfa5460cb/clients/githubrepo/tarball.go#L60-L64

File: clients/azuredevopsrepo/zip.go (hunk 3 above)
Comment (spencerschrock) [reply clients/azuredevopsrepo/zip.go:251]: Did you mean to push a commit before resolving this ?

File: clients/azuredevopsrepo/zip.go (hunk 2 above)
Comment (spencerschrock) [reply clients/azuredevopsrepo/zip.go:183]: optional: You also use `filepath.Clean(z.tempDir) + string(os.PathSeparator)` in a few places in this loop. worth extracting this to one place outside the loop?

```go
destinationPrefix := filepath.Clean(z.tempDir) + string(os.PathSeparator)
for _, file := range zipReader.File {
	//nolint:gosec // G305: Handling of file paths is done below
	filenamepath := filepath.Join(z.tempDir, file.Name)
	if !strings.HasPrefix(filenamepath, destinationPrefix) {
		return errInvalidFilePath
	}
	// ...
	filename := strings.TrimPrefix(filenamepath, destinationPrefix)
}
```

File: clients/azuredevopsrepo/zip.go [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+			return fmt.Errorf("file.Open: %w", err)
+		}
+
+		written, err := io.CopyN(outFile, rc, maxSize)
+		if err != nil && !errors.Is(err, io.EOF) {
+			return fmt.Errorf("%w io.Copy: %w", errZipNotFound, err)
+		}
+		if written >= maxSize {
+			return errFileTooLarge
+		}
```
Comment (spencerschrock) [inline clients/azuredevopsrepo/zip.go:207]: edge case nit: inclusive should be fine?

```go
if written > maxSize {
	return errFileTooLarge
}
```

**Review by spencerschrock — APPROVED** [review]
Thanks for the examples. Just a few minor things.

> They look good. Here's are a couple of examples

Unrelated to this PR, I'm surprised AZDO doesn't allow Scorecard without a token for a public repo. I can access the web GUI just fine and view commits/files. But trying to run without a token (or with an anonymose connection): 

> Error: scorecard.Run: internal error: ListCommits:error during commitsHandler.setup: request for commits failed with TF400813: The user 'aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa' is not authorized to access this resource.

File: clients/azuredevopsrepo/zip.go (hunk 3 above)
Comment (JamieMagee) [reply clients/azuredevopsrepo/zip.go:251]: Ah, sorry. I made the change locally, but didn't push.

**Pull request merged**